### DISCUSSION TITLE
How to tell what dependencies I need for custom loading?

### DISCUSSION THREAD

**Discussion by GreenImp** [josdejong/mathjs#2934] [Q&A]
Hi,

I'm looking at custom loading to reduce the bundle size, as explained here:
* https://mathjs.org/docs/custom_bundling.html#using-just-a-few-functions
* https://mathjs.org/examples/advanced/custom_evaluate_using_import.js.html

However, I'm struggling to figure out which dependencies I need to include.
I'm **only** using the `Math.evaluate()` method in my code, but it obviously calls other method in the background.

I first tried this:

```javascript
import { create, evaluateDependencies } from 'mathjs/lib/esm/number';

const math = create({ evaluateDependencies });
```

But my tests then failed for missing dependencies. I went through, one by one, adding dependencies until my tests passed, but now I have a long list of them, and I'm not sure if I might have missed anything, or included too much.

This is what I've currently got:

```javascript
import {
  create,
  addDependencies,
  ceilDependencies,
  divideDependencies,
  evaluateDependencies,
  floorDependencies,
  maxDependencies,
  minDependencies,
  modDependencies,
  multiplyDependencies,
  powDependencies,
  subtractDependencies,
  unaryMinusDependencies,
} from 'mathjs/lib/esm/number';

const math = create({
  addDependencies,
  ceilDependencies,
  divideDependencies,
  evaluateDependencies,
  floorDependencies,
  maxDependencies,
  minDependencies,
  modDependencies,
  multiplyDependencies,
  powDependencies,
  subtractDependencies,
  unaryMinusDependencies,
});
```

Is there any way of telling if I've missed any dependencies which my tests may have missed, and how do I know which dependencies I need for the `evaluate()` method? Does it depend on what notation I'm trying to evaluate?

For reference, I did find this ticket; https://github.com/josdejong/mathjs/issues/2229
But, although it solved the issue for the OP, it didn't really clarify how to figure it out.

Cheers

**Comment by josdejong** [comment]
That is an interesting one. The parser has a set of operators that it supports like `+` and `*`, and these operators map to a function like `add` and `multiply`. This is not a dependency right now but maybe it would be better to have that as an explicit dependency. 

You can find a list with [all supported operators in the docs](https://mathjs.org/docs/expressions/syntax.html#operators). However that list does not contain the corresponding function names. It maybe easiest to look inside [src/expression/parse.js](https://github.com/josdejong/mathjs/blob/develop/src/expression/parse.js), search for all occurrences of `new OperatorNode(...)`, and list the function names of all operators from there.

Besides that, the expression parser can evaluate functions like `sqrt(4)` and also constants like in `sin(pi/2)`, but only if `sqrt`, `sin`, and `pi` are defined.
